### PR TITLE
Support TSL compilation on Windows with CUDA support.

### PR DESCRIPTION
**Pull request by eaplatanios**
This is following up on #15444. There's still one issue blocking Windows support that I haven't resolved. I've described it [here](https://github.com/openxla/xla/issues/6993#issuecomment-2257844869). Any suggestions/advice on how to proceed for that one would be helpful. cc @hawkinsp @ddunl and also fyi @metab0t.

### CODE REVIEW

File: third_party/tsl/tsl/profiler/lib/nvtx_utils.cc [hunk 1]
```diff
@@ -36,7 +36,7 @@ namespace {
 // nvtxNameOsThreadA:
 // https://nvidia.github.io/NVTX/doxygen/group___r_e_s_o_u_r_c_e___n_a_m_i_n_g.html
 // This convention may not match the one in tsl::Env::GetCurrentThreadId().
-std::optional<uint32_t> GetCurrentThreadId() {
+std::optional<uint32_t> MaybeGetCurrentThreadId() {
```
Comment (eaplatanios) [inline third_party/tsl/tsl/profiler/lib/nvtx_utils.cc:39]: Avoids a name conflict.

File: third_party/tsl/third_party/gpus/cuda/build_defs.bzl.tpl [hunk 2]
```diff
@@ -104,9 +104,16 @@ def if_cuda_newer_than(wanted_ver, if_true, if_false = []):
     wanted_major = int(wanted_ver.split('_')[0])
     wanted_minor = int(wanted_ver.split('_')[1])
 
-    configured_version = "%{cuda_version}"
-    configured_major = int(configured_version.split('.')[0])
-    configured_minor = int(configured_version.split('.')[1])
+    # Strip "64_" which appears in the CUDA version on Windows.
```
Comment (eaplatanios) [inline third_party/tsl/third_party/gpus/cuda/build_defs.bzl.tpl:107]: This is mostly self-explanatory. In Windows, the CUDA version looks something like `64_121`. There are other parts of the build that already handle such version numbers but it was not being handled properly here.

File: third_party/tsl/third_party/gpus/crosstool/windows/msvc_wrapper_for_nvcc.py.tpl [hunk 3]
```diff
@@ -181,6 +181,9 @@ def InvokeNvcc(argv, log=False):
   nvccopts += ['--keep', '--keep-dir', tempdir]
   # Force C++17 dialect (note, everything in just one string!)
   nvccopts += ['--std c++17']
+  # This is so that nvcc does not complain about MSVC or CLANG.
+  nvccopts += ['-allow-unsupported-compiler']
+  nvccopts += ['--expt-extended-lambda', '--expt-relaxed-constexpr']
```
Comment (eaplatanios) [inline third_party/tsl/third_party/gpus/crosstool/windows/msvc_wrapper_for_nvcc.py.tpl:186]: This is necessary but also mirrors settings that are also used in `crosstool_wrapper_driver_is_not_gcc`.

**Comment by ddunl** [comment]
LGTM, this one will require a patch internally I think so I'll try to get to that soon. Thanks!!

**Review by ddunl — APPROVED** [review]

**Comment by eaplatanios** [comment]
Great, thanks @ddunl!

**Comment by vam-google** [comment]
How is this checking that CUDA works on Windows? Is it compilation-specific? If it compiles on windows it does not mean it works on windows

**Pull request closed without merging**